ahciem: Emulate SES enclosure for AHCI enclosures

freqlabs opened the pull request:
Mostly complete.  AHCI controllers supporting enclosure management will
create a virtual SES logical unit on a dedicated SCSI host, allowing
control/status of EM LEDs. A link to the device in a given slot is created for
mapping the slot to a device:
```
root@trunasmini:~# strings /sys/class/enclosure/*/Slot*/device/inquiry 
ATA     Samsung SSD 860 1B6Q
ATA     Samsung SSD 840 6B0Q
ATA     SanDisk SD6SB1M0600 
ATA     SanDisk SDSSDHII00RL
```

Still todo:
 - [x] report a more reasonable serial number than all zeros
 - [x] modify SES driver to support association of ATA drives
 - [x] add a module parameter to allow opting out or forcing attachment
 - [x] correctly check for administratively disabled ports

Jira: https://jira.ixsystems.com/browse/NAS-111023

Commit:
ahciem: Emulate SES enclosure for AHCI enclosures

Mostly complete.  AHCI controllers supporting enclosure management will
create a virtual SES logical unit on a dedicated SCSI host, allowing
control/status of EM LEDs.

Still todo:
 - report a more reasonable serial number than all zeros
 - modify SES driver to support association of ATA drives
 - add a module parameter to allow opting out or forcing attachment

The block device for a drive will not be presented in the enclosure
sysfs until the SES driver supports matching ATA devices.  It currently
only allows for SAS devices to be matched.  As a hacky workaround in the
meantime, the `slot` reported in the enclosure sysfs corresponds to the
unique print id of the ATA device in the slot, which can be mapped to a
block device as follows:

```
/sys/class/ata_port/ata$id/device/host$((id-1))/target$((id-1)):*/*/block/*
```

freqlabs commented on drivers/ata/ahciem.c:
```diff
@@ -0,0 +1,588 @@
+			rbuf[offset + 3] |= 0x10; /* DEVICE OFF */
+
+		/* XXX: potentially out of sync with emp->led_state */
+		memcpy(rbuf + offset, args->enc->status[i], 4);
```
Moved the copy to the top of the loop as it is in FreeBSD.

freqlabs commented:
Whoops I squashed the wrong branch, will update with the correct one.

amotin commented on drivers/ata/ahciem.c:
```diff
@@ -0,0 +1,588 @@
+				  u8 sk, u8 asc, u8 ascq)
+{
+	scsi_build_sense(cmd, false, sk, asc, ascq);
+}
```
Is this obfuscation just to save one parameter?

freqlabs replied:
This is something borrowed from libata. It makes more sense in the 5.10 kernel I am testing with because the convenient `scsi_build_sense` is not available.

amotin commented on drivers/ata/ahciem.c:
```diff
@@ -0,0 +1,588 @@
+	memcpy(rbuf + 32, "2.00", INQUIRY_REVISION_LEN);
+	memcpy(rbuf + 36, "0001", 4);
+	memcpy(rbuf + 40, "S-E-S ", 6);
+	memcpy(rbuf + 46, "2.00", 4);
```
I am trying to recall whether those 3 fields are used by SCSI or only by SEMB.  Can't find them in SES specification.

amotin commented on drivers/ata/ahciem.c:
```diff
@@ -0,0 +1,588 @@
+	memcpy(p, "SGPIO Enclosure                        ", ATA_ID_PROD_LEN);
+	p += ATA_ID_PROD_LEN;
+	/* TODO */
+	memcpy(p, "00000000000000000000", ATA_ID_SERNO_LEN);
```
I'd better not report any identification rather that fixed non-unique.

amotin commented on drivers/ata/ahciem.c:
```diff
@@ -0,0 +1,588 @@
+	struct Scsi_Host *shost;
+	unsigned int id;
+
+	shost = ((struct Scsi_Host *)args->enc) - 1;
```
I think proper way to do this would include container_of().

freqlabs replied:
That's what I originally tried but it didn't build. I'll take another look at the error there.

freqlabs replied:
This part was eliminated.

amotin commented on drivers/ata/ahciem.c:
```diff
@@ -0,0 +1,588 @@
+	/* enclosure logical identifier */
+	memcpy(p, enc_desc, sizeof(enc_desc));
+	p += sizeof(enc_desc);
+	snprintf(p, 7, "ahciem%01u", id); /* XXX: seems fragile */
```
Yes, this ID should be consistent between reboots and preferably between hardware changes.  Number of ATA controller may change if some external is added.

amotin commented on drivers/ata/ahciem.c:
```diff
@@ -0,0 +1,588 @@
+	memcpy(p, "2.00", INQUIRY_REVISION_LEN);
+	p += INQUIRY_REVISION_LEN;
+
+	/* vendor specific enclosure information */
```
It is not a vendor specific enclosure information, it is a type descriptor and following type descriptor tests.

amotin commented on drivers/ata/ahciem.c:
```diff
@@ -0,0 +1,588 @@
+		else if (ata_link_offline(link)) /* XXX: idk if right? */
+			status = ENCLOSURE_STATUS_UNAVAILABLE;
+		else
+			status = ENCLOSURE_STATUS_NOT_INSTALLED;
```
I wonder whether you ever get here now, but ENCLOSURE_STATUS_NOT_INSTALLED is what should be reported for empty enclosure.

freqlabs replied:
One thing that isn't obvious is that `ata_link_online(link)` does not necessarily imply `!ata_link_offline(link)`. Both can return false.

```
/**
 *      ata_link_online - test whether the given link is online
 *      @link: ATA link to test
 *
 *      Test whether @link is online.  This is identical to
 *      ata_phys_link_online() when there's no slave link.  When
 *      there's a slave link, this function should only be called on
 *      the master link and will return true if any of M/S links is
 *      online.
 *
 *      LOCKING:
 *      None.
 *
 *      RETURNS:
 *      True if the port online status is available and online.
 */
 ```
 ```
/**
 *      ata_link_offline - test whether the given link is offline
 *      @link: ATA link to test
 *
 *      Test whether @link is offline.  This is identical to
 *      ata_phys_link_offline() when there's no slave link.  When
 *      there's a slave link, this function should only be called on
 *      the master link and will return true if both M/S links are
 *      offline.
 *
 *      LOCKING:
 *      None.
 *
 *      RETURNS:
 *      True if the port offline status is available and offline.
 */
```

amotin commented on drivers/ata/ahciem.c:
```diff
@@ -0,0 +1,588 @@
+			status = ENCLOSURE_STATUS_NOT_INSTALLED;
+		rbuf[offset] |= status;
+
+		if (ata_link_offline(link)) /* XXX: idk if right? */
```
Probably not. ch->disablephy on FreeBSD meant administratively disabled.

amotin commented on drivers/ata/ahciem.c:
```diff
@@ -0,0 +1,588 @@
+			rbuf[offset] |= 0x80;	/* invalid */
+			continue;
+		}
+		if (sata_pmp_attached(ap) /* XXX: || ch->devices == 0? */)
```
"ch->devices == 0" On FreeBSD means no device probed on the channel.

amotin commented on drivers/ata/ahciem.c:
```diff
@@ -0,0 +1,588 @@
+			rbuf[offset] |= 0x80;	/* invalid */
+
+		/* ATA Element Status (NB: non-standard) */
+		scsi_ulto4b(ap->print_id, rbuf + offset + 4);
```
I'd expect here SCSI bus number for the ATA channel, since SES driver operates in context of SCSI, not ATA.

freqlabs replied:
This is the hack I used to map to a block device for now, since I haven't made the SES driver able to match ATA devices yet.

amotin commented on drivers/ata/ahciem.c:
```diff
@@ -0,0 +1,588 @@
+{
+	ahciem_scsi_set_sense(cmd, ILLEGAL_REQUEST, 0x24, 0x0);
+	scsi_set_sense_field_pointer(cmd->sense_buffer, SCSI_SENSE_BUFFERSIZE,
+				     field, bit, 1);
```
The past argument in scsi_set_sense_field_pointer() specifies whether the error is in command or data, while you've mixed them together.

amotin commented on drivers/ata/ahciem.c:
```diff
@@ -0,0 +1,588 @@
+	if (scsi_sg_copy_to_buffer(cmd, hdr, sizeof(hdr)) != sizeof(hdr)) {
+		ahciem_scsi_set_sense(cmd, ABORTED_COMMAND, 0x34, 0x0);
+		return;
+	}
```
Length of the data transfer should be read from PARAMETER LIST LENGTH of the SEND DIAGNOSTIC command.  You should not need to first read the header to get the length of the further transfer.  Initiator must provide enough data for the specified value, short write unlike read IIRC considered an error.

amotin commented on drivers/ata/ahciem.c:
```diff
@@ -0,0 +1,588 @@
+
+	if (rc == 0)
+		cmd->result = SAM_STAT_GOOD;
+}
```
I don't see you ever use ALLOCATION LENGTH field of the commands to know how much data is requested, neither you track how much data you have.  If initiator requested more data than you have, you should probably somehow return the residual amount.  In some cases initiator may need to know how much data in the received buffer are valid.  Your present code may work, but it is not completely correct.

freqlabs replied:
It works the same way in libata-scsi.c so I'm not sure it's worth correcting.

amotin replied:
I see. But usually fact that somebody is doing worse is not a very good excuse.

amotin reviewed (commented):
The progress looks good, but more detailed comments are inline.

I'd not push it until "modify SES driver to support association of ATA drives" part is implemented, since it does not have much sense to implement the temporary workaround in middleware.

Commit:
address some feedback

Commit:
improve enclosure status reports

Commit:
better ids

Commit:
Return to a sequential counter for unique id

Busses can come and go, attachment order is more reliable.  Use this for
NAA identifiers as well.

Commit:
Properly set field pointer in data

Commit:
Make SES driver match devices in AHCI enclosures

Commit:
Use available BE get/put functions

Commit:
Add ahciem_enable module param to ahci

Commit:
Fix check for disabled device

amotin commented on drivers/scsi/ses.c:
```diff
@@ -650,11 +663,20 @@ static int ses_intf_add(struct device *cdev,
+		if (scsi_is_ata(sdev)) {
+			struct ata_port *ap = ata_shost_to_port(sdev->host);
+			struct ahci_host_priv *hpriv = ap->host->private_data;
+			struct Scsi_Host *shost = hpriv->em_shost;
```
This looks like a gross hack...

freqlabs replied:
`struct ata_host` has an array of `struct ata_port *`, and each port has a corresponding `struct Scsi_Host`. An AHCI controller is a specialization of `struct ata_host` with fields specific to AHCI (including several related to enclosure management) accessed through the host `private_data` field. The enclosure management features of an AHCI controller do not have a corresponding `struct ata_port`, so I have added the field `struct Scsi_Host *em_shost` to the AHCI struct for keeping track of the `struct Scsi_Host` allocated for SES emulation. That way we can go from a `struct scsi_device *sdev` for a device attached to an AHCI controller to the `struct Scsi_Host *shost` for its enclosure management.

What's gross about it? I could maybe clean it up a bit by hiding the details in a libahci function, and I do make the assumption that the ATA port belongs to an AHCI controller which is more of an oversight than a gross hack. I can add that check.

amotin replied:
I see it as huge layering violation for generic SES code to have access to the internal structures of the ATA driver.  I'd prefer it to not even know about the driver used.  My FreeBSD code is maximally opaque, using only numeric bus/target numbers for disk/slot mapping, that if needed could be reused for any other driver without adding more specific conditions, public headers, etc.

freqlabs replied:
Ok I can think about how to improve that, but for now is it ok to merge for nightly builds?

Commit:
Add ahciem support to ahci-platform

Commit:
Check for AHCI rather than assuming

ghost commented:
I'll squash and merge once approved.

amotin commented on drivers/ata/ahci.c:
```diff
@@ -631,6 +631,10 @@ static struct pci_driver ahci_pci_driver = {
 
+static int ahciem_enable = 1;
+module_param(ahciem_enable, int, 0644);
+MODULE_PARM_DESC(ahciem_enable, "Emulate SES enclosure (0 = disabled)");
```
On FreeBSD I recently made it tri-state: 0 -- disable, 1 -- enable, -1 (default) -- enable if hardware supports.  It should allow enclosure management just for physical location purposes on systems without LEDs.  Though I don't insist for now.

amotin commented on drivers/ata/ahci.h:
```diff
@@ -447,4 +448,14 @@ static inline int ahci_nr_ports(u32 cap)
+struct ahciem_enclosure {
+	struct ata_host *host;
+	u8 status[AHCI_MAX_PORTS][4];
+};
```
As I have told before, I'd prefer this to be private.

amotin commented on drivers/ata/ahciem.c:
```diff
@@ -0,0 +1,560 @@
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
```
Below the file is marked as BSD/GPL.  Looking on NTB code as example, here should also be a mention of GPL.

amotin commented on drivers/ata/ahciem.c:
```diff
@@ -0,0 +1,560 @@
+		else if (ata_link_online(link))
+			status = ENCLOSURE_STATUS_OK;
+		else if (ata_link_offline(link))
+			status = ENCLOSURE_STATUS_UNAVAILABLE;
```
I think here should be ENCLOSURE_STATUS_NOT_INSTALLED.

amotin commented on drivers/ata/ahciem.c:
```diff
@@ -0,0 +1,560 @@
+
+		/* ATA Element Status (NB: non-standard) */
+		put_unaligned_be32(i, rbuf + offset + 4);
+		put_unaligned_be32(ap->scsi_host->host_no + 1, rbuf + offset + 8);
```
Why " + 1"?  If you need it to be non-zero in SES, why not hack there?

amotin commented on drivers/scsi/ses.c:
```diff
@@ -766,10 +788,27 @@ static int ses_intf_add(struct device *cdev,
+				continue;
+
+			shost_for_each_device(tmp_sdev, ap->scsi_host) {
+				ses_match_to_enclosure(edev, tmp_sdev, 0);
```
Here it is totally upside-down -- SES already reports you bus numbers for each slot.  You should traverse through the slots and for each one look for the devices.

freqlabs replied:
Oh insightful! That lets me make the ahciem_enclosure structure private again, too!

amotin reviewed (commented):
Here is few more comments. Please fix few obvious and I don't mind it to be merged into nightlies for middleware/UI testing.  Though my objections about layering violations and desired refactoring in SES stay.

Commit:
Fix ahciem license and top comment

Commit:
Keep ahciem_enclosure structure private

Commit:
Fix component removal

Commit:
Isolate addr hack to one file

Commit:
Use ENCLOSURE_STATUS_NOT_INSTALLED

ghost commented:
I've addressed most of the feedback and will now squash and merge this for testing in nightlies. I have started refactoring the SES driver to avoid the layering violation and will merge that later when complete.
